### PR TITLE
Delegate all API calls of gRPC into `opentelemetry_exporter_otlp_grpc_client`, and make it contains all symbols needed.

### LINKED ISSUE #1998: OTLP example fails to build with error "undefined reference to `grpc::Status::OK' "
**Describe your environment** 
WSL ubuntu jammy

**Steps to reproduce**
Install grpc as per doc: https://grpc.io/docs/languages/cpp/quickstart/
Build opentelemetry,
```
cmake .. -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_STANDARD=17 -DBUILD_SHARED_LIBS=ON -DWITH_ABSEIL=OFF -DWITH_OTLP=ON -DBUILD_TESTING=OFF
cmake --build . --target all

```

**What is the expected behavior?**
Build should go through

**What is the actual behavior?**
Got error,
```
[ 79%] Linking CXX executable example_otlp_grpc
cd /home/vivek/Client-Server/module/opentelemetry-cpp/build/examples/otlp && /usr/bin/cmake -E cmake_link_script CMakeFiles/example_otlp_grpc.dir/link.txt --verbose=1
/usr/bin/c++ -rdynamic CMakeFiles/example_otlp_grpc.dir/grpc_main.cc.o -o example_otlp_grpc  -Wl,-rpath,/home/vivek/Client-Server/module/opentelemetry-cpp/build/examples/common/foo_library:/home/vivek/Client-Server/module/opentelemetry-cpp/build/exporters/otlp:/home/vivek/Client-Server/module/opentelemetry-cpp/build/sdk/src/trace:/home/vivek/Client-Server/module/opentelemetry-cpp/build/sdk/src/resource:/home/vivek/Client-Server/module/opentelemetry-cpp/build/sdk/src/metrics:/home/vivek/Client-Server/module/opentelemetry-cpp/build/sdk/src/common ../common/foo_library/libcommon_foo_library.so ../../exporters/otlp/libopentelemetry_exporter_otlp_grpc.so ../../exporters/otlp/libopentelemetry_otlp_recordable.so ../../sdk/src/trace/libopentelemetry_trace.so ../../sdk/src/resource/libopentelemetry_resources.so ../../sdk/src/metrics/libopentelemetry_metrics.so ../../sdk/src/common/libopentelemetry_common.so ../../exporters/otlp/libopentelemetry_exporter_otlp_grpc_client.so ../../libopentelemetry_proto.a /usr/lib/x86_64-linux-gnu/libprotobuf.a
/usr/bin/ld: ../../exporters/otlp/libopentelemetry_exporter_otlp_grpc.so: undefined reference to `grpc::Status::OK'
collect2: error: ld returned 1 exit status
```

**Additional context**
grpc installation looks fine,
```
vivek@LAPTOP-7E1991MU:/usr/lib$ ls | grep "grpc"
libgrpc++.a
libgrpc++_alts.a
libgrpc++_error_details.a
libgrpc++_reflection.a
libgrpc++_unsecure.a
libgrpc.a
libgrpc_authorization_provider.a
libgrpc_plugin_support.a
libgrpc_unsecure.a
libgrpcpp_channelz.a
```

I did the exact steps provided by grpc project, but looks like this build can't link to libgrpc

### PR DESCRIPTION
**Pull request by owent**
Fixes #1998 

## Changes

+ Delegate all API calls of gRPC into `opentelemetry_exporter_otlp_grpc_client`, and make it contains all symbols needed.
+ Add CI job to build shared libraries with static gRPC.
+ Add unit test to check #1603 .

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

File: .vscode/launch.json
Comment (ThomsonTan) [inline .vscode/launch.json:8]: Is this build path from bazel, and need manual update for CMake build?

File: exporters/otlp/BUILD
Comment (ThomsonTan) [inline exporters/otlp/BUILD:271]: Is service proto not needed for grpc exporter?

Comment (owent) [reply exporters/otlp/BUILD:271]: It's already depended by logs_service_grpc_cc below.

File: .vscode/launch.json
Comment (owent) [reply .vscode/launch.json:8]: Sorry.this file is committed by mistake.I will restore it tomorrow.

**Comment by VivekSubr** [comment]
Changes are in only in bazel BUILD, no changes are required in any CMake file?

**Comment by owent** [comment]
> Changes are in only in bazel BUILD, no changes are required in any CMake file?

The main change is moving gRPC callings into `otlp_grpc_client.cc` , and it changes the dependencies of otlp_grpc_client. With cmake, otlp_grpc_client already linked `opentelemetry_proto` before.

**Comment by VivekSubr** [comment]
@owent - cloned your branch and verified it builds, thank you!

**Comment by lalitb** [comment]
@owent Just curious, if you know how was it working in CI env, and also our local setup. As I could see opentelemetry_exporter_otlp_grpc_client  having gRPC++ symbols locally in WSL ubuntu 22.04. OR is it anything to do with compiler toolchain in customer environment ?

**Comment by owent** [comment]
> @owent Just curious, if you know how was it working in CI env, and also our local setup. As I could see opentelemetry_exporter_otlp_grpc_client having gRPC++ symbols locally in WSL ubuntu 22.04. OR is it anything to do with compiler toolchain in customer environment ?

This problem happens only when we built gRPC as static libraries and build otel-cpp as dynamic libraries.In CI jobs, we only build both gRPC and otel-cpp as static or dynamic libraries. Do you think we should add a jobs to check this situation?

File: exporters/otlp/src/otlp_grpc_client.cc
Comment (lalitb) [inline exporters/otlp/src/otlp_grpc_client.cc:106]: where is this method being used ?

**Comment by lalitb** [comment]
> This problem happens only when we built gRPC as static libraries and build otel-cpp as dynamic libraries.

Thanks @owent . This is strange, as I normally use this setup locally on my machine ( i.e., linking dynamic otel-cpp libraries with static gRPC) but never had this error :) But if @VivekSubr is validating the problem is solved with the PR, it should be fine :)

**Comment by owent** [comment]
Sorry but some commits are missing in this PR, I will push them again on next Monday.

**Comment by owent** [comment]
All commits are pushed now.

**Comment by lalitb** [comment]
Sorry for the delay in reviewing this. The changes look good in general, I just want to test these changes as I always had the working setup of dynamic otel-cpp libraries with static gRPC. Will be completing the review in this week.

**Comment by owent** [comment]
> Sorry for the delay in reviewing this. The changes look good in general, I just want to test these changes as I always have the working setup of dynamic otel-cpp libraries with static gRPC. Will be completing the review in this week.

This PR add a ci job `cmake_otprotocol_shared_libs_with_static_grpc_test` to test it.

**Review by marcalff — APPROVED** [review]
Looks good to me.

My only question is about `OtlpGrpcClient::MakeCompletionQueue()`,
as already mentioned by @lalitb 

Please merge with a recent main, to keep the PR up to date, and check CI.

File: exporters/otlp/src/otlp_grpc_client.cc
Comment (owent) [reply exporters/otlp/src/otlp_grpc_client.cc:106]: I plan use it to implement concurrency exporting in the future. Also , the `grpc::CompletionQueue` is also used by `Export` .I think we better export it.

**Comment by owent** [comment]
> Looks good to me.
> 
> My only question is about `OtlpGrpcClient::MakeCompletionQueue()`, as already mentioned by @lalitb
> 
> Please merge with a recent main, to keep the PR up to date, and check CI.

Sorry , I missed the comment before. I plan use it to implement concurrency exporting in the future. Also , the `grpc::CompletionQueue` is also used by `Export` of gRPC services. I think we better export it.

**Comment by marcalff** [comment]
> Sorry , I missed the comment before. I plan use it to implement concurrency exporting in the future. Also , the `grpc::CompletionQueue` is also used by `Export` of gRPC services. I think we better export it.

Ok, approved.

**Comment by marcalff** [comment]
@lalitb @ThomsonTan  - Any chance this PR can go in 1.9.0 ?

**Comment by lalitb** [comment]
> @lalitb @ThomsonTan - Any chance this PR can go in 1.9.0 ?

Looking into it now. Sorry for delay, was on vacation, back today morning.

**Comment by ThomsonTan** [comment]
Do we plan to include this in v1.9.0 or after that? Probably merge it after the release because it is not a trivial change.

**Comment by lalitb** [comment]
> Do we plan to include this in v1.9.0 or after that? Probably merge it after the release because it is not a trivial change.

I agree. While changes look good going through the PR, and has been open for long time, still let's not merge it at last moment before the release. Taking lesson from previous releases where we introduced breaking changes with merges just before the release.

**Comment by marcalff** [comment]
> > Do we plan to include this in v1.9.0 or after that? Probably merge it after the release because it is not a trivial change.
> 
> I agree. While changes look good going through the PR, and has been open for long time, still let's not merge it at last moment before the release. Taking lesson from previous releases where we introduced breaking changes with merges just before the release.

Ok, will merge after 1.9.0 then.

**Review by lalitb — APPROVED** [review]

**Pull request merged**
